Tiled scene 2d renderer

nyalldawson opened the pull request:
(temporarily includes https://github.com/qgis/QGIS/pull/54166 )

Implements 2d renderers for tiled scene layers. We currently offer three renderer choices:

- Textured (default). Loads the texture image and paints the mesh using these textures

![image](https://github.com/qgis/QGIS/assets/1829991/a470b408-2fc5-47aa-a098-8efe95d81553)

- Texture color. Uses the center point color for rendering each mesh triangle. Gives a clear visual overview of the mesh structure while also providing some context via the texture color

![image](https://github.com/qgis/QGIS/assets/1829991/c0cf1dca-0d14-40f9-83b2-46e90695d897)

- Mesh. Renders the scene's mesh using a fill symbol

![image](https://github.com/qgis/QGIS/assets/1829991/4f7be11d-49e7-4359-9a4b-78f25ed08d77)

nyalldawson committed:
Implement 2d renderer for tiled scenes

We currently offer three renderer choices:
- Textured (default). Loads the texture image and paints the mesh using
  these textures
- Texture color. Uses the center point color for rendering each mesh
  triangle. Gives a clear visual overview of the mesh structure while
  also providing some context via the texture color
- Mesh. Renders the scene's mesh using a fill symbol

nyalldawson committed:
Const correctness

nyalldawson committed:
Add option to show boundaries

nyalldawson committed:
Const

wonder-sk commented on src/core/tiledscene/qgstiledscenelayerrenderer.cpp:
```diff
@@ -75,51 +90,550 @@ bool QgsTiledSceneLayerRenderer::render()
+    if ( !tile.isValid() || !tileIsVisibleInMap( tile ) )
+      continue;
+
+    switch ( mIndex.childAvailability( tileId ) )
```
is all this necessary? If the request did not have the `NoHierarchyFetch` flag set, I would have expected that all relevant children should be available?

nyalldawson replied:
It's designed this way so that we can apply the extra short cut logic from https://github.com/qgis/QGIS/blob/2a913c4efa77babd8c2ef2101c218a7057212c65/src/core/tiledscene/qgstiledscenelayerrenderer.cpp#L159 .

Otherwise we end up requesting data from outside the visible map extent in a lot of cases (verified!).

wonder-sk commented on src/core/tiledscene/qgstiledscenerendererregistry.cpp:
```diff
@@ -17,13 +17,21 @@
+               QObject::tr( "Texture Color" ),
+               QgsTiledSceneTextureColorRenderer::create ) );
+  addRenderer( new QgsTiledSceneRendererMetadata( QStringLiteral( "mesh" ),
+               QObject::tr( "Mesh" ),
```
Shall we cal this renderer "wireframe" ? It seems to me that term is more commonly used, and we would have fewer meaning associated with the "mesh" term...

nyalldawson replied:
Good point :+1: 

I'm planning on adding rendering of line primitives there too, but a "wireframe" name still applies even with that!

wonder-sk commented on src/core/tiledscene/qgstiledscenetexturecolorrenderer.cpp:
```diff
@@ -0,0 +1,87 @@
+  painter->setPen( Qt::NoPen );
+  painter->setBrush( b );
+  painter->drawPolygon( triangle );
+}
```
Is the texture color renderer much faster than the "true" texture renderer? I like the artistic effect of this renderer, but I am wondering if it will be generally useful - any situations when someone would want to use this one instead of full texture renderer?

nyalldawson replied:
From the PR description: "Gives a clear visual overview of the mesh structure while also providing some context via the texture color". It's kind of a trade off between the full texture view (which doesn't reveal the data structure) vs the complete loss of coloring information which the mesh/wireframe renderer gives.

I was toying with just making this a checkbox option in the texture renderer though. Or alternatively it could be a option for the wireframe renderer, so that users could color their individually crafted fill symbol with the representative color...

nyalldawson replied:
I've moved this option to the wireframe renderer

nyalldawson commented on src/core/tiledscene/qgstiledscenelayerrenderer.cpp:
```diff
@@ -75,51 +90,550 @@ bool QgsTiledSceneLayerRenderer::render()
+                                  QgsPointXY( mapExtent.xMaximum(), mapYCenter )
+                                );
+  const double mapMetersPerPixel = mapWidthMeters / context->outputSize().width();
+  const double maximumErrorInMeters = maximumErrorPixels * mapMetersPerPixel;
```
> that's assuming that MUPP is in meters, if not, multiply by a factor

That's the issue I ran into -- because we set scene layers to a geographic crs, if you just load in a scene layer initially then your project will be in geographic units, and there's no meaningful map unit -> meter conversion we can do. 

We could potentially bypass this with some (more!) logic to only use QgsDistanceArea when the map units are geographic... but that's just throwing more code in here :laughing:

nyalldawson commented on src/core/tiledscene/qgstiledscenelayerrenderer.h:
```diff
@@ -68,9 +112,32 @@ class CORE_EXPORT QgsTiledSceneLayerRenderer: public QgsMapLayerRenderer
+      QPolygonF triangle;
+      double z;
+      QPair< int, int > textureId { -1, -1 };
+      double textureCoords[6];
```
Done

nyalldawson committed:
Use float for texture coordinates

nyalldawson committed:
Fix clang tidy warning

nyalldawson committed:
Rename mesh renderer to wireframe renderer

nyalldawson committed:
Correctly set default fill symbol for wireframe renderer

nyalldawson committed:
Remove texture color renderer, and replace with checkbox option
in wireframe renderer

nyalldawson committed:
Reduce artifacts in textured tiled scene renderer by slightly
buffering triangles before rendering, to avoid visible
gaps between neighbouring triangles

nyalldawson committed:
Add sample tiled scene

nyalldawson committed:
Add unit tests
